// @create-figma-plugin/ui/lib/components/layer/stories/layer-component-unselected.stories.js
import { h } from 'preact';
import { useState } from 'preact/hooks';
import { useInitialFocus } from '../../../hooks/use-initial-focus/use-initial-focus';
import { IconLayerComponent16 } from '../../../icons/icon-16/icon-layer-component-16';
import { Layer } from '../layer';
export default {
    parameters: {
        fixedWidth: true,
        order: 3
    },
    title: "Components/Layer/Component Unselected"
};
export const Passive = function () {
    const [value, setValue] = useState(false);
    function handleChange(event) {
        const newValue = event.currentTarget.checked;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Layer, { component: true, icon: h(IconLayerComponent16, null), onChange: handleChange, value: value }, "Text"));
};
export const Focused = function () {
    const [value, setValue] = useState(false);
    function handleChange(event) {
        const newValue = event.currentTarget.checked;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Layer, { ...useInitialFocus(), component: true, icon: h(IconLayerComponent16, null), onChange: handleChange, value: value }, "Text"));
};
export const Bold = function () {
    const [value, setValue] = useState(false);
    function handleChange(event) {
        const newValue = event.currentTarget.checked;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Layer, { bold: true, component: true, icon: h(IconLayerComponent16, null), onChange: handleChange, value: value }, "Text"));
};
export const Description = function () {
    const [value, setValue] = useState(false);
    function handleChange(event) {
        const newValue = event.currentTarget.checked;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Layer, { component: true, description: "Description", icon: h(IconLayerComponent16, null), onChange: handleChange, value: value }, "Text"));
};
export const LongText = function () {
    const longText = Array(100).fill("Text").join(" ");
    const [value, setValue] = useState(false);
    function handleChange(event) {
        const newValue = event.currentTarget.checked;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Layer, { component: true, icon: h(IconLayerComponent16, null), onChange: handleChange, value: value }, longText));
};
export const LongTextDescription = function () {
    const longText = Array(100).fill("Text").join(" ");
    const description = Array(100).fill("Text").join(" ");
    const [value, setValue] = useState(false);
    function handleChange(event) {
        const newValue = event.currentTarget.checked;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Layer, { component: true, description: description, icon: h(IconLayerComponent16, null), onChange: handleChange, value: value }, longText));
};
export const OnValueChange = function () {
    const longText = Array(100).fill("Text").join(" ");
    const [value, setValue] = useState(false);
    function handleValueChange(newValue) {
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Layer, { component: true, icon: h(IconLayerComponent16, null), onValueChange: handleValueChange, value: value }, longText));
};
