// @create-figma-plugin/ui/lib/components/dropdown/stories/dropdown-default.stories.js
import { Fragment, h } from 'preact';
import { useState } from 'preact/hooks';
import { useInitialFocus } from '../../../hooks/use-initial-focus/use-initial-focus';
import { IconLayerFrame16 } from '../../../icons/icon-16/icon-layer-frame-16';
import { VerticalSpace } from '../../../layout/vertical-space/vertical-space';
import { Button } from '../../button/button';
import { Dropdown } from '../dropdown';
export default {
    parameters: {
        fixedWidth: true,
        order: 1
    },
    title: "Components/Dropdown/Default"
};
export const Empty = function () {
    const [value, setValue] = useState(null);
    const options = [
        { value: "foo" },
        { value: "bar" },
        { value: "baz" },
        { separator: true },
        { header: "Header" },
        { value: "qux" }
    ];
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(Dropdown, { onChange: handleChange, options: options, value: value });
};
export const EmptyManyOptions = function () {
    const [value, setValue] = useState(null);
    const options = [...Array(100).keys()].map(function (index) {
        return { value: `${index + 1}` };
    });
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(Dropdown, { onChange: handleChange, options: options, value: value });
};
export const Placeholder = function () {
    const [value, setValue] = useState(null);
    const options = [
        { value: "foo" },
        { value: "bar" },
        { value: "baz" },
        { separator: true },
        { header: "Header" },
        { value: "qux" }
    ];
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Dropdown, { onChange: handleChange, options: options, placeholder: "placeholder", value: value }));
};
export const Filled = function () {
    const [value, setValue] = useState("foo");
    const options = [
        { value: "foo" },
        { value: "bar" },
        { value: "baz" },
        { separator: true },
        { header: "Header" },
        { value: "qux" }
    ];
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(Dropdown, { onChange: handleChange, options: options, value: value });
};
export const FilledManyOptions = function () {
    const [value, setValue] = useState("42");
    const options = [...Array(100).keys()].map(function (index) {
        return { value: `${index + 1}` };
    });
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(Dropdown, { onChange: handleChange, options: options, value: value });
};
export const LongText = function () {
    const [value, setValue] = useState(Array(20).fill("foo").join(" "));
    const options = [
        { value: Array(20).fill("foo").join(" ") },
        { value: Array(20).fill("bar").join(" ") },
        { value: Array(20).fill("baz").join(" ") },
        { separator: true },
        { header: "Header" },
        { value: Array(4).fill("qux").join(" ") }
    ];
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(Dropdown, { onChange: handleChange, options: options, value: value });
};
export const Focused = function () {
    const [value, setValue] = useState("foo");
    const options = [
        { value: "foo" },
        { value: "bar" },
        { value: "baz" },
        { separator: true },
        { header: "Header" },
        { value: "qux" }
    ];
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Dropdown, { ...useInitialFocus(), onChange: handleChange, options: options, value: value }));
};
export const Disabled = function () {
    const options = [
        { value: "foo" },
        { value: "bar" },
        { value: "baz" },
        { separator: true },
        { header: "Header" },
        { value: "qux" }
    ];
    function handleChange() {
        throw new Error("This function should not be called");
    }
    return (h(Dropdown, { disabled: true, onChange: handleChange, options: options, value: "foo" }));
};
export const OptionDisabled = function () {
    const [value, setValue] = useState("foo");
    const options = [
        { value: "foo" },
        { disabled: true, value: "bar" },
        { value: "baz" },
        { separator: true },
        { header: "Header" },
        { value: "qux" }
    ];
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(Dropdown, { onChange: handleChange, options: options, value: value });
};
export const OptionText = function () {
    const [value, setValue] = useState("a");
    const options = [
        {
            text: "foo",
            value: "a"
        },
        { text: "bar", value: "b" },
        { text: "baz", value: "c" },
        { separator: true },
        { header: "Header" },
        { text: "qux", value: "d" }
    ];
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(Dropdown, { onChange: handleChange, options: options, value: value });
};
export const Icon = function () {
    const [value, setValue] = useState("foo");
    const options = [
        { value: "foo" },
        { value: "bar" },
        { value: "baz" },
        { separator: true },
        { header: "Header" },
        { value: "qux" }
    ];
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Dropdown, { icon: h(IconLayerFrame16, null), onChange: handleChange, options: options, value: value }));
};
export const IconDisabled = function () {
    const [value, setValue] = useState("foo");
    const options = [
        { value: "foo" },
        { value: "bar" },
        { value: "baz" },
        { separator: true },
        { header: "Header" },
        { value: "qux" }
    ];
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Dropdown, { disabled: true, icon: h(IconLayerFrame16, null), onChange: handleChange, options: options, value: value }));
};
export const IconText = function () {
    const [value, setValue] = useState("foo");
    const options = [
        { value: "foo" },
        { value: "bar" },
        { value: "baz" },
        { separator: true },
        { header: "Header" },
        { value: "qux" }
    ];
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Dropdown, { icon: "X", onChange: handleChange, options: options, value: value }));
};
export const IconTextDisabled = function () {
    const [value, setValue] = useState("foo");
    const options = [
        { value: "foo" },
        { value: "bar" },
        { value: "baz" },
        { separator: true },
        { header: "Header" },
        { value: "qux" }
    ];
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Dropdown, { disabled: true, icon: "X", onChange: handleChange, options: options, value: value }));
};
export const DynamicallySetOptions = function () {
    const [value, setValue] = useState(null);
    const [options, setOptions] = useState([
        { value: "foo" }
    ]);
    function handleClick() {
        setOptions([
            { value: "foo" },
            { value: "bar" },
            { value: "baz" },
            { separator: true },
            { header: "Header" },
            { value: "qux" }
        ]);
    }
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Fragment, null, h(Button, { onClick: handleClick }, "Set"), h(VerticalSpace, { space: "small" }), h(Dropdown, { onChange: handleChange, options: options, placeholder: "placeholder", value: value })));
};
export const FlexParent = function () {
    const [valueA, setValueA] = useState(Array(20).fill("foo").join(" "));
    const optionsA = [
        { value: Array(20).fill("foo").join(" ") },
        { value: "bar" }
    ];
    function handleChangeA(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValueA(newValue);
    }
    const [valueB, setValueB] = useState("qux");
    const optionsB = [
        { value: Array(20).fill("baz").join(" ") },
        { value: "qux" }
    ];
    function handleChangeB(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValueB(newValue);
    }
    const style = { display: "flex" };
    return (h("div", { style: style }, h(Dropdown, { onChange: handleChangeA, options: optionsA, value: valueA }), ")", h(Dropdown, { onChange: handleChangeB, options: optionsB, value: valueB }), ")"));
};
export const OnValueChange = function () {
    const [value, setValue] = useState("foo");
    const options = [
        { value: "foo" },
        { value: "bar" },
        { value: "baz" },
        { separator: true },
        { header: "Header" },
        { value: "qux" }
    ];
    function handleValueChange(newValue) {
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Dropdown, { onValueChange: handleValueChange, options: options, value: value }));
};
export const TopLeft = function () {
    const [value, setValue] = useState(null);
    const options = [
        { value: "foo" },
        { value: "bar" },
        { value: "baz" },
        { separator: true },
        { header: "Header" },
        { value: "qux" }
    ];
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    const style = {
        left: "-24px",
        position: "absolute",
        top: "-12px",
        width: "240px"
    };
    return (h("div", { style: style }, h(Dropdown, { onChange: handleChange, options: options, value: value })));
};
export const TopLeftManyOptions = function () {
    const [value, setValue] = useState(null);
    const options = [...Array(100).keys()].map(function (index) {
        return { value: `${index + 1}` };
    });
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    const style = {
        left: "-24px",
        position: "absolute",
        top: "-12px",
        width: "240px"
    };
    return (h("div", { style: style }, h(Dropdown, { onChange: handleChange, options: options, value: value })));
};
export const TopRight = function () {
    const [value, setValue] = useState(null);
    const options = [
        { value: "foo" },
        { value: "bar" },
        { value: "baz" },
        { separator: true },
        { header: "Header" },
        { value: "qux" }
    ];
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    const style = {
        position: "absolute",
        right: "-24px",
        top: "-12px",
        width: "240px"
    };
    return (h("div", { style: style }, h(Dropdown, { onChange: handleChange, options: options, value: value })));
};
export const TopRightManyOptions = function () {
    const [value, setValue] = useState(null);
    const options = [...Array(100).keys()].map(function (index) {
        return { value: `${index + 1}` };
    });
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    const style = {
        position: "absolute",
        right: "-24px",
        top: "-12px",
        width: "240px"
    };
    return (h("div", { style: style }, h(Dropdown, { onChange: handleChange, options: options, value: value })));
};
export const MiddleLeft = function () {
    const [value, setValue] = useState(null);
    const options = [
        { value: "foo" },
        { value: "bar" },
        { value: "baz" },
        { separator: true },
        { header: "Header" },
        { value: "qux" }
    ];
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    const style = {
        left: "-24px",
        position: "absolute",
        top: "50%",
        transform: "translateY(-50%)",
        width: "240px"
    };
    return (h("div", { style: style }, h(Dropdown, { onChange: handleChange, options: options, value: value })));
};
export const MiddleLeftManyOptions = function () {
    const [value, setValue] = useState(null);
    const options = [...Array(100).keys()].map(function (index) {
        return { value: `${index + 1}` };
    });
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    const style = {
        left: "-24px",
        position: "absolute",
        top: "50%",
        transform: "translateY(-50%)",
        width: "240px"
    };
    return (h("div", { style: style }, h(Dropdown, { onChange: handleChange, options: options, value: value })));
};
export const MiddleRight = function () {
    const [value, setValue] = useState(null);
    const options = [
        { value: "foo" },
        { value: "bar" },
        { value: "baz" },
        { separator: true },
        { header: "Header" },
        { value: "qux" }
    ];
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    const style = {
        position: "absolute",
        right: "-24px",
        top: "50%",
        transform: "translateY(-50%)",
        width: "240px"
    };
    return (h("div", { style: style }, h(Dropdown, { onChange: handleChange, options: options, value: value })));
};
export const MiddleRightManyOptions = function () {
    const [value, setValue] = useState(null);
    const options = [...Array(100).keys()].map(function (index) {
        return { value: `${index + 1}` };
    });
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    const style = {
        position: "absolute",
        right: "-24px",
        top: "50%",
        transform: "translateY(-50%)",
        width: "240px"
    };
    return (h("div", { style: style }, h(Dropdown, { onChange: handleChange, options: options, value: value })));
};
export const BottomLeft = function () {
    const [value, setValue] = useState(null);
    const options = [
        { value: "foo" },
        { value: "bar" },
        { value: "baz" },
        { separator: true },
        { header: "Header" },
        { value: "qux" }
    ];
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    const style = {
        bottom: "-12px",
        left: "-24px",
        position: "absolute",
        width: "240px"
    };
    return (h("div", { style: style }, h(Dropdown, { onChange: handleChange, options: options, value: value })));
};
export const BottomLeftManyOptions = function () {
    const [value, setValue] = useState(null);
    const options = [...Array(100).keys()].map(function (index) {
        return { value: `${index + 1}` };
    });
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    const style = {
        bottom: "-12px",
        left: "-24px",
        position: "absolute",
        width: "240px"
    };
    return (h("div", { style: style }, h(Dropdown, { onChange: handleChange, options: options, value: value })));
};
export const BottomRight = function () {
    const [value, setValue] = useState(null);
    const options = [
        { value: "foo" },
        { value: "bar" },
        { value: "baz" },
        { separator: true },
        { header: "Header" },
        { value: "qux" }
    ];
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    const style = {
        bottom: "-12px",
        position: "absolute",
        right: "-24px",
        width: "240px"
    };
    return (h("div", { style: style }, h(Dropdown, { onChange: handleChange, options: options, value: value })));
};
export const BottomRightManyOptions = function () {
    const [value, setValue] = useState(null);
    const options = [...Array(100).keys()].map(function (index) {
        return { value: `${index + 1}` };
    });
    function handleChange(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    const style = {
        bottom: "-12px",
        position: "absolute",
        right: "-24px",
        width: "240px"
    };
    return (h("div", { style: style }, h(Dropdown, { onChange: handleChange, options: options, value: value })));
};
