// @create-figma-plugin/ui/lib/components/textbox/textbox/stories/textbox-default.stories.js
import { MIXED_STRING } from '@create-figma-plugin/utilities';
import { h } from 'preact';
import { useState } from 'preact/hooks';
import { useInitialFocus } from '../../../../hooks/use-initial-focus/use-initial-focus';
import { IconLayerFrame16 } from '../../../../icons/icon-16/icon-layer-frame-16';
import { Textbox } from '../textbox';
export default {
    parameters: {
        fixedWidth: true,
        order: 1
    },
    title: "Components/Textbox/Default"
};
export const Empty = function () {
    const [value, setValue] = useState("");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(Textbox, { onInput: handleInput, value: value });
};
export const EmptyFocused = function () {
    const [value, setValue] = useState("");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(Textbox, { ...useInitialFocus(), onInput: handleInput, value: value });
};
export const Placeholder = function () {
    const [value, setValue] = useState("");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Textbox, { onInput: handleInput, placeholder: "placeholder", value: value }));
};
export const PlaceholderFocused = function () {
    const [value, setValue] = useState("");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Textbox, { ...useInitialFocus(), onInput: handleInput, placeholder: "placeholder", value: value }));
};
export const Filled = function () {
    const [value, setValue] = useState("Text");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(Textbox, { onInput: handleInput, value: value });
};
export const Focused = function () {
    const [value, setValue] = useState("Text");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(Textbox, { ...useInitialFocus(), onInput: handleInput, value: value });
};
export const Disabled = function () {
    function handleInput() {
        throw new Error("This function should not be called");
    }
    return h(Textbox, { disabled: true, onInput: handleInput, value: "Text" });
};
export const Icon = function () {
    const [value, setValue] = useState("Text");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Textbox, { icon: h(IconLayerFrame16, null), onInput: handleInput, value: value }));
};
export const IconDisabled = function () {
    const [value, setValue] = useState("Text");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Textbox, { disabled: true, icon: h(IconLayerFrame16, null), onInput: handleInput, value: value }));
};
export const IconText = function () {
    const [value, setValue] = useState("Text");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(Textbox, { icon: "X", onInput: handleInput, value: value });
};
export const IconTextDisabled = function () {
    const [value, setValue] = useState("Text");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(Textbox, { disabled: true, icon: "X", onInput: handleInput, value: value });
};
export const Mixed = function () {
    const [value, setValue] = useState(MIXED_STRING);
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(Textbox, { onInput: handleInput, value: value });
};
export const RevertOnEscapeKeyDown = function () {
    const [value, setValue] = useState("Text");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(Textbox, { onInput: handleInput, revertOnEscapeKeyDown: true, value: value });
};
export const ValidateOnBlur = function () {
    const [value, setValue] = useState("Text");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    function validateOnBlur(value) {
        return value !== "";
    }
    return (h(Textbox, { onInput: handleInput, validateOnBlur: validateOnBlur, value: value }));
};
export const Password = function () {
    const [value, setValue] = useState("Text");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(Textbox, { onInput: handleInput, password: true, value: value });
};
export const OnValueInput = function () {
    const [value, setValue] = useState("Text");
    function handleValueInput(newValue) {
        console.log(newValue);
        setValue(newValue);
    }
    return h(Textbox, { onValueInput: handleValueInput, value: value });
};
