// @create-figma-plugin/ui/lib/components/textbox/textbox-numeric/stories/textbox-numeric-default.stories.js
import { MIXED_STRING } from '@create-figma-plugin/utilities';
import { h } from 'preact';
import { useState } from 'preact/hooks';
import { useInitialFocus } from '../../../../hooks/use-initial-focus/use-initial-focus';
import { IconLayerFrame16 } from '../../../../icons/icon-16/icon-layer-frame-16';
import { TextboxNumeric } from '../textbox-numeric';
export default {
    parameters: {
        fixedWidth: true,
        order: 1
    },
    title: "Components/Textbox Numeric/Default"
};
export const Empty = function () {
    const [value, setValue] = useState("");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(TextboxNumeric, { onInput: handleInput, value: value });
};
export const EmptyFocused = function () {
    const [value, setValue] = useState("");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(TextboxNumeric, { ...useInitialFocus(), onInput: handleInput, value: value }));
};
export const Placeholder = function () {
    const [value, setValue] = useState("");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(TextboxNumeric, { onInput: handleInput, placeholder: "placeholder", value: value }));
};
export const PlaceholderFocused = function () {
    const [value, setValue] = useState("");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(TextboxNumeric, { ...useInitialFocus(), onInput: handleInput, placeholder: "placeholder", value: value }));
};
export const Filled = function () {
    const [value, setValue] = useState("42");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(TextboxNumeric, { onInput: handleInput, value: value });
};
export const Focused = function () {
    const [value, setValue] = useState("42");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(TextboxNumeric, { ...useInitialFocus(), onInput: handleInput, value: value }));
};
export const Disabled = function () {
    function handleInput() {
        throw new Error("This function should not be called");
    }
    return h(TextboxNumeric, { disabled: true, onInput: handleInput, value: "42" });
};
export const Icon = function () {
    const [value, setValue] = useState("42");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(TextboxNumeric, { icon: h(IconLayerFrame16, null), onInput: handleInput, value: value }));
};
export const IconDisabled = function () {
    const [value, setValue] = useState("42");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(TextboxNumeric, { disabled: true, icon: h(IconLayerFrame16, null), onInput: handleInput, value: value }));
};
export const IconText = function () {
    const [value, setValue] = useState("42");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(TextboxNumeric, { icon: "X", onInput: handleInput, value: value });
};
export const IconTextDisabled = function () {
    const [value, setValue] = useState("42");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(TextboxNumeric, { disabled: true, icon: "X", onInput: handleInput, value: value }));
};
export const Mixed = function () {
    const [value, setValue] = useState(MIXED_STRING);
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(TextboxNumeric, { onInput: handleInput, value: value });
};
export const RevertOnEscapeKeyDown = function () {
    const [value, setValue] = useState("42");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(TextboxNumeric, { onInput: handleInput, revertOnEscapeKeyDown: true, value: value }));
};
export const ValidateOnBlur = function () {
    const [value, setValue] = useState("42");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    function validateOnBlur(value) {
        console.log(value);
        return value !== null;
    }
    return (h(TextboxNumeric, { onInput: handleInput, validateOnBlur: validateOnBlur, value: value }));
};
export const IntegersOnly = function () {
    const [value, setValue] = useState("42");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(TextboxNumeric, { integer: true, onInput: handleInput, value: value });
};
export const CustomIncrements = function () {
    const [value, setValue] = useState("42");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(TextboxNumeric, { incrementBig: 8, incrementSmall: 4, onInput: handleInput, value: value }));
};
export const CustomSmallIncrements = function () {
    const [value, setValue] = useState("0");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(TextboxNumeric, { incrementBig: 0.5, incrementSmall: 0.1, onInput: handleInput, value: value }));
};
export const MinimumMaximum = function () {
    const [value, setValue] = useState("0");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(TextboxNumeric, { maximum: 10, minimum: 0, onInput: handleInput, value: value }));
};
export const Suffix = function () {
    const [value, setValue] = useState("100%");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(TextboxNumeric, { maximum: 100, minimum: 0, onInput: handleInput, suffix: "%", value: value }));
};
export const SuffixMixed = function () {
    const [value, setValue] = useState(MIXED_STRING);
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return h(TextboxNumeric, { onInput: handleInput, suffix: "%", value: value });
};
export const SuffixRevertOnEscapeKeyDown = function () {
    const [value, setValue] = useState("100%");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(TextboxNumeric, { maximum: 100, minimum: 0, onInput: handleInput, revertOnEscapeKeyDown: true, suffix: "%", value: value }));
};
export const SuffixValidateOnBlur = function () {
    const [value, setValue] = useState("100%");
    function handleInput(event) {
        const newValue = event.currentTarget.value;
        console.log(newValue);
        setValue(newValue);
    }
    function validateOnBlur(value) {
        if (value === null) {
            return 100;
        }
        return true;
    }
    return (h(TextboxNumeric, { maximum: 100, minimum: 0, onInput: handleInput, revertOnEscapeKeyDown: true, suffix: "%", validateOnBlur: validateOnBlur, value: value }));
};
export const OnValueInput = function () {
    const [value, setValue] = useState("42");
    function handleValueInput(newValue) {
        console.log(newValue);
        setValue(newValue);
    }
    function handleNumericValueInput(newNumericValue) {
        console.log(newNumericValue);
    }
    return (h(TextboxNumeric, { onNumericValueInput: handleNumericValueInput, onValueInput: handleValueInput, value: value }));
};
