// @create-figma-plugin/ui/lib/components/toggle/stories/toggle-selected.stories.js
import { h } from 'preact';
import { useState } from 'preact/hooks';
import { useInitialFocus } from '../../../hooks/use-initial-focus/use-initial-focus';
import { Text } from '../../text/text';
import { Toggle } from '../toggle';
export default {
    parameters: {
        order: 2
    },
    title: "Components/Toggle/Selected"
};
export const Passive = function () {
    const [value, setValue] = useState(true);
    function handleChange(event) {
        const newValue = event.currentTarget.checked;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Toggle, { onChange: handleChange, value: value }, h(Text, null, "Text")));
};
export const Focused = function () {
    const [value, setValue] = useState(true);
    function handleChange(event) {
        const newValue = event.currentTarget.checked;
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Toggle, { ...useInitialFocus(), onChange: handleChange, value: value }, h(Text, null, "Text")));
};
export const Disabled = function () {
    function handleChange() {
        throw new Error("This function should not be called");
    }
    return (h(Toggle, { disabled: true, onChange: handleChange, value: true }, h(Text, null, "Text")));
};
export const OnValueChange = function () {
    const [value, setValue] = useState(true);
    function handleValueChange(newValue) {
        console.log(newValue);
        setValue(newValue);
    }
    return (h(Toggle, { onValueChange: handleValueChange, value: value }, h(Text, null, "Text")));
};
