// @create-figma-plugin/ui/lib/components/button/stories/button-danger.stories.js
import { h } from 'preact';
import { useInitialFocus } from '../../../hooks/use-initial-focus/use-initial-focus';
import { Button } from '../button';
export default {
    parameters: {
        order: 2
    },
    title: "Components/Button/Danger"
};
export const Passive = function () {
    function handleClick(event) {
        console.log(event);
    }
    return (h(Button, { danger: true, onClick: handleClick }, "Text"));
};
export const Focused = function () {
    function handleClick(event) {
        console.log(event);
    }
    return (h(Button, { ...useInitialFocus(), danger: true, onClick: handleClick }, "Text"));
};
export const Disabled = function () {
    function handleClick() {
        throw new Error("This function should not be called");
    }
    return (h(Button, { danger: true, disabled: true, onClick: handleClick }, "Text"));
};
export const Loading = function () {
    function handleClick() {
        throw new Error("This function should not be called");
    }
    return (h(Button, { danger: true, loading: true, onClick: handleClick }, "Text"));
};
export const LoadingFocused = function () {
    function handleClick() {
        throw new Error("This function should not be called");
    }
    return (h(Button, { ...useInitialFocus(), danger: true, loading: true, onClick: handleClick }, "Text"));
};
export const LoadingDisabled = function () {
    function handleClick() {
        throw new Error("This function should not be called");
    }
    return (h(Button, { danger: true, disabled: true, loading: true, onClick: handleClick }, "Text"));
};
export const FullWidth = function () {
    function handleClick(event) {
        console.log(event);
    }
    return (h(Button, { danger: true, fullWidth: true, onClick: handleClick }, "Text"));
};
FullWidth.parameters = {
    fixedWidth: true
};
export const FullWidthLongText = function () {
    const longText = Array(100).fill("Text").join(" ");
    function handleClick(event) {
        console.log(event);
    }
    return (h(Button, { danger: true, fullWidth: true, onClick: handleClick }, longText));
};
FullWidthLongText.parameters = {
    fixedWidth: true
};
